File: app/src/main/java/com/ash/studios/musify/Adapters/PlaylistAdapter.java
package com.ash.studios.musify.Adapters;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ash.studios.musify.Activities.Categories.BunchList;
import com.ash.studios.musify.Model.Playlist;
import com.ash.studios.musify.Model.Song;
import com.ash.studios.musify.R;
import com.ash.studios.musify.Utils.Utils;
import com.bumptech.glide.Glide;

import java.util.ArrayList;

public class PlaylistAdapter extends RecyclerView.Adapter<PlaylistAdapter.VH> {
    private Context context;
    private ArrayList<Playlist> list;

    public PlaylistAdapter(Context context, ArrayList<Playlist> list, ProgressBar pb, TextView nf) {
        this.context = context;
        this.list = list;
        if (pb != null) pb.setVisibility(View.GONE);
        if (nf != null && getItemCount() == 0) nf.setVisibility(View.VISIBLE);
    }

    @NonNull
    @Override
    public PlaylistAdapter.VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new VH(LayoutInflater.from(context).inflate(R.layout.item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull PlaylistAdapter.VH holder, int position) {
        Playlist playlist = list.get(position);
        ArrayList<Song> songs = playlist.getSongs();

        holder.playlistArtist.setVisibility(View.GONE);
        holder.playlistName.setText(playlist.getName());
        holder.playlistCount.setText(new StringBuilder("\u266B ").append(songs.size()));
        if (playlist.getSongs().size() > 0)
            Glide.with(context.getApplicationContext())
                    .asBitmap().load(Utils.getAlbumArt(playlist.getSongs().get(0).getAlbum_id()))
                    .placeholder(R.mipmap.icon)
                    .into(holder.playlistCover);

        holder.itemView.setOnClickListener(v -> context.startActivity(new Intent(context, BunchList.class)
                .putExtra("list_from", "Playlists")
                .putExtra("list_name", playlist.getName())
                .putExtra("list", songs)));
    }

    @Override
    public int getItemCount() {
        return list == null ? 0 : list.size();
    }

    static class VH extends RecyclerView.ViewHolder {
        TextView playlistName, playlistArtist, playlistCount;
        ImageView playlistCover;

        public VH(@NonNull View itemView) {
            super(itemView);

            playlistName = itemView.findViewById(R.id.title);
            playlistArtist = itemView.findViewById(R.id.artist);
            playlistCount = itemView.findViewById(R.id.duration);
            playlistCover = itemView.findViewById(R.id.album_art);
        }
    }
}

File: app/src/main/java/com/ash/studios/musify/adapters/PlaylistAdapter.kt
package com.ash.studios.musify.adapters

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.ash.studios.musify.Models.Playlist
import com.ash.studios.musify.R
import com.ash.studios.musify.activities.categories.PLBunchList
import com.ash.studios.musify.activities.categories.PlayList
import com.ash.studios.musify.utils.Utils
import com.bumptech.glide.Glide
import com.futuremind.recyclerviewfastscroll.SectionTitleProvider

class PlaylistAdapter(private val context: Context, private val list: ArrayList<Playlist>?, pb: ProgressBar?, nf: TextView?) : RecyclerView.Adapter<PlaylistAdapter.VH>(), SectionTitleProvider {
    init {
        if (pb != null) pb.visibility = View.GONE
        if (nf != null && itemCount == 0) nf.visibility = View.VISIBLE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        return VH(LayoutInflater.from(context).inflate(R.layout.item, parent, false))
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        val playlist = list!![position]
        val songs = playlist.songs
        holder.playlistArtist.visibility = View.GONE
        holder.playlistName.text = playlist.name
        holder.playlistCount.text = StringBuilder("\u266B ").append(songs.size)
        if (playlist.songs.size > 0) Glide.with(context.applicationContext)
            .asBitmap().load(Utils.getAlbumArt(playlist.songs[0].album_id))
            .placeholder(R.drawable.placeholder)
            .into(holder.playlistCover)
        holder.itemView.setOnClickListener {
            if (playlist.songs.size > 0) context.startActivity(
                Intent(context, PLBunchList::class.java)
                    .putExtra("position", position)
                    .putExtra("playlist", playlist)
            ) else Toast.makeText(context, "No songs in that playlist :(", Toast.LENGTH_SHORT).show()
        }
        holder.itemView.setOnLongClickListener {
            val dialog = Utils.getDialog(context, R.layout.delete_dg)
            val title = dialog.findViewById<TextView>(R.id.del_dg_title)
            val body = dialog.findViewById<TextView>(R.id.del_dg_body)
            val cancel = dialog.findViewById<TextView>(R.id.close_del_dg_btn)
            val delete = dialog.findViewById<TextView>(R.id.del_song_btn)
            title.text = StringBuilder("Remove playlist?")
            body.text = StringBuilder("Selected playlist will be removed from this playlists")
            cancel.setOnClickListener { dialog.dismiss() }
            delete.setOnClickListener {
                dialog.dismiss()
                list.removeAt(position)
                notifyItemRemoved(position)
                notifyItemRangeChanged(position, itemCount)
                Utils.deletePlaylist(context, playlist)
                if (itemCount == 0) {
                    (context as PlayList).finish()
                    Toast.makeText(context, "No playlist found.", Toast.LENGTH_SHORT).show()
                }
            }
            true
        }
    }

    override fun getItemCount(): Int {
        return list?.size ?: 0
    }

    override fun getSectionTitle(position: Int): String {
        return list!![position].name.substring(0, 1)
    }

    class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var playlistName: TextView
        var playlistArtist: TextView
        var playlistCount: TextView
        var playlistCover: ImageView

        init {
            playlistName = itemView.findViewById(R.id.title)
            playlistArtist = itemView.findViewById(R.id.artist)
            playlistCount = itemView.findViewById(R.id.duration)
            playlistCover = itemView.findViewById(R.id.album_art)
        }
    }
}